bpo-37000: Remove obsolete comment in _randbelow_with_getrandbits

matthiasgoergens opened the pull request:
In `random._randbelow_with_getrandbits` we are generating a random integer
in [0,n).  We do that by repeatedly drawing a random number, r, in [0,2**k)
until we hit an r < n.

That's a great strategy.  It's not only very simple, but even with worst
case input, we only need to draw 2 random numbers on average, before we
hit the jackpot.

However, the code is a bit overly cautious.  When n is a power of 2, we
ask `getrandbits` to sample from [0,2*n].  The code justifies that
strange behaviour with a comment expressing fear of the special case n==1.

That fear is unfounded: the obvious code works just fine for n==1.

Fixing this not only makes the code simpler to understand, it also
guarantees that the important special case of n being a power of two now
has a guaranteed worst case runtime of only a single call to
getrandbits; instead of an expected runtime of two calls with no worst
case bound.

Existing tests already cover this code and the special cases of
interest.  So we don't need any new tests.

This minor performance bug was introduced in
0515661314c4e5b9235e07b2c46b8f456c7fadc3 as far as I can tell.  But I
can't really tell what the original author was thinking.

Since this is a fairly trivial change, I did not create an issue.  (But I am happy to create one, if you think it's better this way.)

I discovered this problem while working on https://discuss.python.org/t/random-choice-from-a-dict/17834

Automerge-Triggered-By: GH:rhettinger

matthiasgoergens committed:
Fix corner case in _randbelow_with_getrandbits

In random._randbelow_with_getrandbits we are generating a random integer
in [0,n).  We do that by repeatedly drawing a random number, r, in [0,2**k)
until we hit an r < n.

That's a great strategy.  It's not only very simple, but even with worst
case input, we only need to draw 2 random numbers on average, before we
hit the jackpot.

However, the code is a bit overly cautious.  When n is a power of 2, we
ask `getrandbits` to sample from [0,2*n].  The code justifies that
strange behaviour with a comment expressing fear of the special case n==1.

That fear is unfounded: the obvious code works just fine for n==1.

Fixing this not only makes the code simpler to understand, it also
guarantees that the important special case of n being a power of two now
has a guaranteed worst case runtime of only a single call to
getrandbits; instead of an expected runtime of two calls with no worst
case bound.

Existing tests already cover this code and the special cases of
interest.  So we don't need any new tests.

This minor performance bug was introduced in
0515661314c4e5b9235e07b2c46b8f456c7fadc3 as far as I can tell.  But I
can't really tell what the original author was thinking.

matthiasgoergens committed:
Revert "Fix corner case in _randbelow_with_getrandbits"

This reverts commit f2a869491c86a98f1e9b416a62f0833a10658a38.

matthiasgoergens committed:
Explain why we can't have nice things

rhettinger commented on Lib/random.py:
```diff
@@ -239,7 +239,13 @@ def _randbelow_with_getrandbits(self, n):
 
         getrandbits = self.getrandbits
-        k = n.bit_length()  # don't use (n-1) here because n can be 1
+        # For efficiency, we'd like to use k = (n-1).bit_length() here, but
```
Let's just remove the comment entirely.

matthiasgoergens replied:
That sounds like an improvement over the status quo of having a misleading comment to me.

I guess we can stick the more in-depth explanation in the commit description, so at least it shows up in git blame?

matthiasgoergens committed:
Remove explanatory comment

For efficiency, we'd like to use k = (n-1).bit_length() here, but
before Python 3.9, `getrandbits()` did not accept 0 as an argument.
That's why using `n-1` was a problem when `n` was 1.

Now we are stuck with this version, because we want to reproduce
results after explicitly setting a seed even between releases.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

rhettinger committed:
Delete 2022-08-08-10-57-32.gh-issue-13485.xtWoE8.rst

Comment changes don't warrant a comment.
